File: trading-algo/src/shared/events.ts
import type { EventType, TradingEvent } from './types.js';

type EventHandler<T = unknown> = (event: TradingEvent<T>) => void | Promise<void>;

/**
 * Event bus for inter-module communication.
 * All team members communicate through events, keeping them decoupled.
 */
export class EventBus {
  private handlers = new Map<EventType, Set<EventHandler>>();
  private allHandlers = new Set<EventHandler>();

  on<T = unknown>(type: EventType, handler: EventHandler<T>): () => void {
    if (!this.handlers.has(type)) {
      this.handlers.set(type, new Set());
    }
    this.handlers.get(type)!.add(handler as EventHandler);
    return () => this.handlers.get(type)?.delete(handler as EventHandler);
  }

  onAll(handler: EventHandler): () => void {
    this.allHandlers.add(handler);
    return () => this.allHandlers.delete(handler);
  }

  async emit<T = unknown>(type: EventType, data: T, source: string): Promise<void> {
    const event: TradingEvent<T> = {
      type,
      data,
      timestamp: Date.now(),
      source,
    };

    const handlers = this.handlers.get(type) ?? new Set();
    const promises: Promise<void>[] = [];

    for (const handler of handlers) {
      const result = handler(event);
      if (result instanceof Promise) promises.push(result);
    }
    for (const handler of this.allHandlers) {
      const result = handler(event);
      if (result instanceof Promise) promises.push(result);
    }

    await Promise.allSettled(promises);
  }

  clear(): void {
    this.handlers.clear();
    this.allHandlers.clear();
  }
}

export const eventBus = new EventBus();

File: trading-algo-py/shared/events.py
"""Event bus for decoupled inter-module communication."""

from __future__ import annotations

import logging
import time
from collections import defaultdict
from typing import Any, Callable

log = logging.getLogger(__name__)

EventHandler = Callable[..., Any]


class EventBus:
    """Simple pub/sub event bus."""

    def __init__(self) -> None:
        self._handlers: dict[str, list[EventHandler]] = defaultdict(list)

    def on(self, event_type: str, handler: EventHandler) -> None:
        self._handlers[event_type].append(handler)

    def off(self, event_type: str, handler: EventHandler) -> None:
        if handler in self._handlers[event_type]:
            self._handlers[event_type].remove(handler)

    def emit(self, event_type: str, data: Any = None, source: str = "") -> None:
        handlers = self._handlers.get(event_type, [])
        for handler in handlers:
            try:
                handler(data, source)
            except Exception as e:
                log.error("Event handler error for %s: %s", event_type, e)


event_bus = EventBus()
